File: Queue/649. Dota2 Senate/solution.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> R;
        queue<int> D;

        for (int i=0; i<senate.size(); i++) {
            if (senate[i] == 'R') {
                R.push(i);
            } else {
                D.push(i);
            }
        }

        // ban round by round
        int n = senate.size();
        while (!R.empty() && !D.empty()) {
            int i = R.front();
            R.pop();
            int j = D.front();
            D.pop();

            if (i < j) {
                // R ban D
                R.push(i+n); // keep it in queue because R still can be banned by R
            } else {
                // D ban R
                D.push(j+n);
            }
        }

        return R.empty() ? "Dire" : "Radiant";
    }
};

File: Queue/649. Dota2 Senate/solution.py
from collections import deque

class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        # round-based procedure
        # until last one win
        
        ban = defaultdict(int)
        nxt = deque()
        cur = deque(senate)
        exists = {}
        while cur:
            while cur:
                party = cur.popleft()
                if party == "R" and ban["R"] > 0:
                    ban["R"] -= 1
                elif party == "D" and ban["D"] > 0:
                    ban["D"] -= 1
                else:
                    if party == "R":
                        ban["D"] += 1
                    else:
                        ban["R"] += 1

                    nxt.append(party)
                    exists[party] = True
            if len(exists) == 1:
                return "Radiant" if "R" in exists else "Dire"
            cur, nxt = nxt, deque()
            exists = {}

class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        R = deque()
        D = deque()
        n = len(senate)

        for i in range(n):
            if senate[i] == 'R':
                R.append(i)
            else:
                D.append(i)

        while R and D:
            r_idx = R.popleft()
            d_idx = D.popleft()

            if r_idx < d_idx:
                R.append(r_idx + n)
            else:
                D.append(d_idx + n)

        return "Radiant" if R else "Dire"